osu.Android/OsuGameActivity.cs

```diff
@@ -23,8 +32,65 @@ protected override void OnCreate(Bundle savedInstanceState)
 
             base.OnCreate(savedInstanceState);
 
+            // OnNewIntent() only fires for an activity if it's *re-launched* while it's on top of the activity stack.
+            // on first launch we still have to fire manually.
+            // reference: https://developer.android.com/reference/android/app/Activity#onNewIntent(android.content.Intent)
+            handleIntent(Intent);
+
             Window.AddFlags(WindowManagerFlags.Fullscreen);
             Window.AddFlags(WindowManagerFlags.KeepScreenOn);
         }
+
+        protected override void OnNewIntent(Intent intent) => handleIntent(intent);
+
+        private void handleIntent(Intent intent)
+        {
+            switch (intent.Action)
+            {
+                case Intent.ActionDefault:
+                    if (intent.Scheme == ContentResolver.SchemeContent)
+                        handleImportFromUri(intent.Data);
+                    break;
+
+                case Intent.ActionSend:
+                {
+                    var content = intent.ClipData?.GetItemAt(0);
+                    if (content != null)
+                        handleImportFromUri(content.Uri);
+                    break;
+                }
+            }
+        }
+
+        private void handleImportFromUri(Uri uri)
+        {
+            // there are more performant overloads of this method, but this one is the most backwards-compatible
+            // (dates back to API 1).
+            var cursor = ContentResolver?.Query(uri, null, null, null, null);
+
+            if (cursor == null)
+                return;
+
+            cursor.MoveToFirst();
+
+            var filenameColumn = cursor.GetColumnIndex(OpenableColumns.DisplayName);
+
+            var stream = ContentResolver.OpenInputStream(uri);
+            string filename = cursor.GetString(filenameColumn);
+
+            if (stream != null)
+                Task.Factory.StartNew(() => runImport(stream, filename), TaskCreationOptions.LongRunning);
+        }
+
+        private Task runImport(Stream stream, string filename)
```

Comment:
thinking we want to dispose the original stream here (after copying it out)? probably safest to.

Reply:
Yeah that's a fair point. I'll grab that later.

Reply:
I think I've resolved this, but not in the most immediate way - apparently [it's inadvisable to do `ContentProvider` stuff on the main UI thread](https://developer.android.com/guide/topics/providers/content-provider-basics#SimpleQuery) so the whole rigamarole is on a background task now.